src/libsyntax/ext/expand.rs

```diff
@@ -188,12 +189,30 @@ impl Invocation {
 
 pub struct MacroExpander<'a, 'b:'a> {
     pub cx: &'a mut ExtCtxt<'b>,
+    partial_expansions: HashMap<Mark, PartialExpansion>,
+    full_expansions: HashMap<Mark, FullExpansion>,
     monotonic: bool, // c.f. `cx.monotonic_expander()`
 }
 
+struct PartialExpansion {
```

Comment:
Could you add a comment here please - it would be good to explain how partial expansions come about - are they just intermediate state or are they stored like this, in what circumstances is a macro partially rather than fully expanded, and are there invariants that can be assumed about partial expansions.